SignalServiceKit/src/Util/Cryptography.m

```diff
@@ -386,137 +385,166 @@ + (NSData *)encryptAttachmentData:(NSData *)attachmentData
+
+    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+    if (!ctx) {
+        OWSFail(@"%@ failed to build context while encrypting", self.tag);
         return nil;
     }
-    
-    uint8_t *authTagBytes = malloc(kAESGCM128_TagLength);
-    if (authTagBytes == NULL) {
-        free(cipherTextBytes);
-        OWSFail(@"%@ Failed to allocate authTagBytes", self.tag);
+
+    // Initialise the encryption operation.
+    if (EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL) != kOpenSSLSuccess) {
+        OWSFail(@"%@ failed to init encryption", self.tag);
         return nil;
     }
-    
-    // NOTE: Since `tagLength` is an input parameter, it seems weird that the signature for tagLength is a `size_t*` rather than just a `size_t`.
-    //
-    // I found a vague reference in the Safari repository implying that this may be a bug:
-    // source: https://www.mail-archive.com/webkit-changes@lists.webkit.org/msg114561.html
-    //
-    // Comment was:
-    //     tagLength is actual an input <rdar://problem/30660074>
-    size_t tagLength = kAESGCM128_TagLength;
-
-    CCCryptorStatus status
-        = CCCryptorGCM(kCCEncrypt,       // CCOperation op, /* kCCEncrypt, kCCDecrypt */
-            kCCAlgorithmAES128,          // CCAlgorithm alg,
-            key.keyData.bytes,           // const void *key, /* raw key material */
-            key.keyData.length,          // size_t keyLength,
-            initializationVector.bytes,  // const void *iv,
-            initializationVector.length, // size_t ivLen,
-            NULL,                        // const void *aData,
-            0,                           // size_t aDataLen,
-            plainTextData.bytes,         // const void *dataIn,
-            plainTextData.length,        // size_t dataInLength,
-            cipherTextBytes,             // void *dataOut,
-            authTagBytes,                // const void *tag,
-            &tagLength                   // size_t *tagLength)
-            );
-
-    if (status != kCCSuccess) {
-        OWSFail(@"CCCryptorGCM encrypt failed with status: %d", status);
-        free(cipherTextBytes);
-        free(authTagBytes);
+
+    // Set IV length if default 12 bytes (96 bits) is not appropriate
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, (int)initializationVector.length, NULL) != kOpenSSLSuccess) {
+        OWSFail(@"%@ failed to set IV length", self.tag);
         return nil;
     }
-    
-    // build up return value: initializationVector || cipherText || authTag
+
+    // Initialise key and IV
+    if (EVP_EncryptInit_ex(ctx, NULL, NULL, key.keyData.bytes, initializationVector.bytes) != kOpenSSLSuccess) {
+        OWSFail(@"%@ failed to set key and iv while encrypting", self.tag);
+        return nil;
+    }
+
+    int bytesEncrypted = 0;
+    // Provide the message to be encrypted, and obtain the encrypted output.
+    // EVP_EncryptUpdate can be called multiple times if necessary
```

Comment:
Ditto: what does this mean: `EVP_EncryptUpdate can be called multiple times if necessary`?